### PR TITLE
[EXPORTER] OTLP file exporter

### LINKED ISSUE #2482: OTLP file exporter
**Is your feature request related to a problem?**

According to 
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/file-exporter.md#file-storage-requirements .

It we have file exporter, some projects will not depend openssl, curl, or gRPC any more.

**Describe the solution you'd like**

Use OTLP utils to convert data to JSON format, just like it in OTLP HTTP exporter.

**Additional context**

There are some questions I want to discuss before implementing it. 
In practice, we usually have log rotate to control the disk usage, but the specification does not say anything about it.
Can I add some extensions to it?
For example, if we configure the file pattern as `%Y-%m-%d/otlp.%N.log` and set the rotate size to three, we want to create a directory every day and rotate output file name as `otlp.0.log`, `otlp.1.log` and `otlp.2.log`. And we may also set alias name `%Y-%m-%d/otlp.log` so the `otlp.log` will always point to the latest one?

### PR DESCRIPTION
**Pull request by owent**
Fixes #2482 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by owent** [comment]
This PR is ready to be reviewed now. @open-telemetry/cpp-approvers @perhapsmaple

File: exporters/otlp/src/otlp_file_client.cc
Comment (perhapsmaple) [inline exporters/otlp/src/otlp_file_client.cc:1539]: I think we should add the force flush mechanism for the std::ostream case as well. In case the user calls `TracerProvider::ForceFlush`, we should add something like this:

```
std::ostream &out = nostd::get<std::reference_wrapper<std::ostream>>(options_.backend_options);
out.flush();
```

Comment (owent) [reply exporters/otlp/src/otlp_file_client.cc:1539]: Good point. I didn't implement it before because some ostreams is not thread-safety and we must ignore `timeout`.
It may be more important to keep it right. I will add flush for ostream later.

**Comment by owent** [comment]
> Hi @owent, thanks for doing this. This is a very nice and well done PR. I have a couple of questions for now:
> 
> 1. I'm not very familiar with this rolling strategy. Most log tailers like Promtail and Fluentbit recommend the Rename and Create rotation strategy. Has this rotation strategy been tested with any existing log tailer? See here for more information: https://grafana.com/docs/loki/latest/send-data/promtail/logrotation/
> 2. I know this is probably a bit out of scope of the opentelemetry specification, but is it possible to make a base class like `FileBackend` so that users can implement their own rotation strategies if required. The `OtlpFileBackend` is very well designed and configurable, but might not fit everyone's use case.
> 
> We will have to test it with the `otlpjsonfilereceiver` from `opentelemetry-collector-contrib` to make sure that everything works properly. I'm a little busy for the next couple days but I can help you with this after.
> 
> Thanks

Thanks.

1. I want to keep the default implementation simple. This rotation strategy is tested in our internal collector agent which is also based on otel-collector.
2. It's a good idea to leave a base class to let users to use their own log sink, maybe spdlog, boost.log or other components.I will add it later. You can review it anytime when you have time. Thanks.

File: CMakeLists.txt
Comment (marcalff) [inline CMakeLists.txt:209]:
```suggestion
option(WITH_OTLP_FILE
       "Whether to include the OTLP file exporter in the SDK" OFF)
```

File: exporters/otlp/src/otlp_file_client.cc
Comment (marcalff) [inline exporters/otlp/src/otlp_file_client.cc:1276]: Do we really want to print these instructions in the log ?

How about just printing the `res` return code, and have these instructions as C++ comments.

File: exporters/otlp/src/otlp_file_client.cc
Comment (marcalff) [inline exporters/otlp/src/otlp_file_client.cc:1279]: Same for non windows:

Do we really want to print these instructions in the log ?

How about just printing the `res` return code, and have these instructions as C++ comments.

**Review by marcalff — APPROVED** [review]
Very good work, as we are used to.

See some minor changes or comments.

**Comment by marcalff** [comment]
Need to resolve my 3 comments, and need a second reviewer: adding the please-review label.

File: CMakeLists.txt
Comment (owent) [reply CMakeLists.txt]: Thaanks and done.

**Comment by owent** [comment]
> Need to resolve my 3 comments, and need a second reviewer: adding the please-review label.

Done

File: CMakeLists.txt
Comment (ThomsonTan) [inline CMakeLists.txt:208]: nit: add a newline after the option.

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_file_client.h
Comment (ThomsonTan) [inline exporters/otlp/include/opentelemetry/exporters/otlp/otlp_file_client.h:15]: This is not needed as it is already declared in the otlp_file_client_options.h included at line 9

**Review by ThomsonTan — APPROVED** [review]

**Comment by lalitb** [comment]
Nice work. Can we document this exporter as an section in [README.md](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters/otlp) ?

**Comment by owent** [comment]
> Nice work. Can we document this exporter as an section in [README.md](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters/otlp) ?

Sorry for the missing. I will add documents tomorrow.

**Comment by marcalff** [comment]
> > Nice work. Can we document this exporter as an section in [README.md](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters/otlp) ?
> 
> Sorry for the missing. I will add documents ~tomorrow~ several days later. Sorry, I'm too busy recently.I will commit and push documents together with codes if there are more feedbacks about this PR.

Merging this PR now, please open a PR later for some doc when you have time.

Thanks for this work.

**Pull request merged**
